File: bindings/typescript/src/safe_ml.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

export interface Result<T> {
  ok: boolean;
  value?: T;
  error?: string;
}

/**
 * Softmax activation function.
 */
export function softmax(values: number[]): Result<number[]> {
  if (values.length === 0) {
    return { ok: false, error: 'Cannot compute softmax of empty array' };
  }

  // Find max for numerical stability
  const max = Math.max(...values);
  const exps = values.map((v) => Math.exp(v - max));
  const sum = exps.reduce((a, b) => a + b, 0);

  if (sum === 0 || !Number.isFinite(sum)) {
    return { ok: false, error: 'Softmax computation overflow' };
  }

  return { ok: true, value: exps.map((e) => e / sum) };
}

/**
 * Sigmoid activation function.
 */
export function sigmoid(x: number): number {
  if (x >= 0) {
    return 1 / (1 + Math.exp(-x));
  } else {
    const exp = Math.exp(x);
    return exp / (1 + exp);
  }
}

/**
 * Sigmoid derivative.
 */
export function sigmoidDerivative(x: number): number {
  const s = sigmoid(x);
  return s * (1 - s);
}

/**
 * ReLU activation function.
 */
export function relu(x: number): number {
  return Math.max(0, x);
}

/**
 * ReLU derivative.
 */
export function reluDerivative(x: number): number {
  return x > 0 ? 1 : 0;
}

/**
 * Leaky ReLU activation.
 */
export function leakyRelu(x: number, alpha: number = 0.01): number {
  return x >= 0 ? x : alpha * x;
}

/**
 * Leaky ReLU derivative.
 */
export function leakyReluDerivative(x: number, alpha: number = 0.01): number {
  return x >= 0 ? 1 : alpha;
}

/**
 * Tanh activation (hyperbolic tangent).
 */
export function tanh(x: number): number {
  return Math.tanh(x);
}

/**
 * Tanh derivative.
 */
export function tanhDerivative(x: number): number {
  const t = Math.tanh(x);
  return 1 - t * t;
}

/**
 * GELU activation (Gaussian Error Linear Unit).
 */
export function gelu(x: number): number {
  return 0.5 * x * (1 + Math.tanh(Math.sqrt(2 / Math.PI) * (x + 0.044715 * Math.pow(x, 3))));
}

/**
 * Swish activation (x * sigmoid(x)).
 */
export function swish(x: number): number {
  return x * sigmoid(x);
}

/**
 * ELU activation (Exponential Linear Unit).
 */
export function elu(x: number, alpha: number = 1.0): number {
  return x >= 0 ? x : alpha * (Math.exp(x) - 1);
}

/**
 * Cross-entropy loss for classification.
 */
export function crossEntropyLoss(predicted: number[], target: number[]): Result<number> {
  if (predicted.length !== target.length) {
    return { ok: false, error: 'Array lengths must match' };
  }
  if (predicted.length === 0) {
    return { ok: false, error: 'Arrays cannot be empty' };
  }

  let loss = 0;
  const epsilon = 1e-15;

  for (let i = 0; i < predicted.length; i++) {
    const p = Math.max(epsilon, Math.min(1 - epsilon, predicted[i]));
    const t = target[i];
    loss -= t * Math.log(p) + (1 - t) * Math.log(1 - p);
  }

  return { ok: true, value: loss / predicted.length };
}

/**
 * Mean squared error loss.
 */
export function mseLoss(predicted: number[], target: number[]): Result<number> {
  if (predicted.length !== target.length) {
    return { ok: false, error: 'Array lengths must match' };
  }
  if (predicted.length === 0) {
    return { ok: false, error: 'Arrays cannot be empty' };
  }

  let sum = 0;
  for (let i = 0; i < predicted.length; i++) {
    const diff = predicted[i] - target[i];
    sum += diff * diff;
  }

  return { ok: true, value: sum / predicted.length };
}

/**
 * Mean absolute error loss.
 */
export function maeLoss(predicted: number[], target: number[]): Result<number> {
  if (predicted.length !== target.length) {
    return { ok: false, error: 'Array lengths must match' };
  }
  if (predicted.length === 0) {
    return { ok: false, error: 'Arrays cannot be empty' };
  }

  let sum = 0;
  for (let i = 0; i < predicted.length; i++) {
    sum += Math.abs(predicted[i] - target[i]);
  }

  return { ok: true, value: sum / predicted.length };
}

/**
 * Huber loss (smooth L1).
 */
export function huberLoss(predicted: number[], target: number[], delta: number = 1.0): Result<number> {
  if (predicted.length !== target.length) {
    return { ok: false, error: 'Array lengths must match' };
  }
  if (predicted.length === 0) {
    return { ok: false, error: 'Arrays cannot be empty' };
  }

  let sum = 0;
  for (let i = 0; i < predicted.length; i++) {
    const error = Math.abs(predicted[i] - target[i]);
    if (error <= delta) {
      sum += 0.5 * error * error;
    } else {
      sum += delta * (error - 0.5 * delta);
    }
  }

  return { ok: true, value: sum / predicted.length };
}

/**
 * Accuracy metric for classification.
 */
export function accuracy(predicted: number[], target: number[]): Result<number> {
  if (predicted.length !== target.length) {
    return { ok: false, error: 'Array lengths must match' };
  }
  if (predicted.length === 0) {
    return { ok: false, error: 'Arrays cannot be empty' };
  }

  let correct = 0;
  for (let i = 0; i < predicted.length; i++) {
    if (Math.round(predicted[i]) === Math.round(target[i])) {
      correct++;
    }
  }

  return { ok: true, value: correct / predicted.length };
}

/**
 * Batch normalization.
 */
export function batchNorm(
  values: number[],
  gamma: number = 1,
  beta: number = 0,
  epsilon: number = 1e-5
): Result<number[]> {
  if (values.length === 0) {
    return { ok: false, error: 'Cannot normalize empty array' };
  }

  const mean = values.reduce((a, b) => a + b, 0) / values.length;
  const variance = values.reduce((sum, v) => sum + (v - mean) ** 2, 0) / values.length;
  const std = Math.sqrt(variance + epsilon);

  const normalized = values.map((v) => gamma * ((v - mean) / std) + beta);

  return { ok: true, value: normalized };
}

/**
 * Layer normalization.
 */
export function layerNorm(
  values: number[],
  gamma: number = 1,
  beta: number = 0,
  epsilon: number = 1e-5
): Result<number[]> {
  return batchNorm(values, gamma, beta, epsilon);
}

/**
 * Dropout (training mode).
 */
export function dropout(values: number[], rate: number = 0.5): number[] {
  if (rate <= 0) return values;
  if (rate >= 1) return new Array(values.length).fill(0);

  const scale = 1 / (1 - rate);
  return values.map((v) => (Math.random() > rate ? v * scale : 0));
}

/**
 * L2 regularization penalty.
 */
export function l2Regularization(weights: number[], lambda: number = 0.01): number {
  return lambda * weights.reduce((sum, w) => sum + w * w, 0);
}

/**
 * L1 regularization penalty.
 */
export function l1Regularization(weights: number[], lambda: number = 0.01): number {
  return lambda * weights.reduce((sum, w) => sum + Math.abs(w), 0);
}

/**
 * Clip gradients to prevent exploding gradients.
 */
export function clipGradients(gradients: number[], maxNorm: number): number[] {
  const norm = Math.sqrt(gradients.reduce((sum, g) => sum + g * g, 0));
  if (norm <= maxNorm) return gradients;
  const scale = maxNorm / norm;
  return gradients.map((g) => g * scale);
}

export const SafeML = {
  // Activations
  softmax,
  sigmoid,
  sigmoidDerivative,
  relu,
  reluDerivative,
  leakyRelu,
  leakyReluDerivative,
  tanh,
  tanhDerivative,
  gelu,
  swish,
  elu,
  // Loss functions
  crossEntropyLoss,
  mseLoss,
  maeLoss,
  huberLoss,
  // Metrics
  accuracy,
  // Normalization
  batchNorm,
  layerNorm,
  dropout,
  // Regularization
  l1Regularization,
  l2Regularization,
  clipGradients,
};

File: bindings/kotlin/src/main/kotlin/com/hyperpolymath/proven/SafeML.kt
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

import kotlin.math.*

/**
 * Machine learning utility functions.
 */
object SafeML {
    // ========== Activation Functions ==========

    /**
     * Sigmoid activation: 1 / (1 + e^-x)
     */
    fun sigmoid(x: Double): Double = 1.0 / (1.0 + exp(-x))

    /**
     * Sigmoid derivative: sigmoid(x) * (1 - sigmoid(x))
     */
    fun sigmoidDerivative(x: Double): Double {
        val s = sigmoid(x)
        return s * (1 - s)
    }

    /**
     * Hyperbolic tangent activation.
     */
    fun tanh(x: Double): Double = kotlin.math.tanh(x)

    /**
     * Tanh derivative: 1 - tanh²(x)
     */
    fun tanhDerivative(x: Double): Double {
        val t = tanh(x)
        return 1 - t * t
    }

    /**
     * ReLU activation: max(0, x)
     */
    fun relu(x: Double): Double = maxOf(0.0, x)

    /**
     * ReLU derivative.
     */
    fun reluDerivative(x: Double): Double = if (x > 0) 1.0 else 0.0

    /**
     * Leaky ReLU activation.
     */
    fun leakyRelu(x: Double, alpha: Double = 0.01): Double = if (x > 0) x else alpha * x

    /**
     * Leaky ReLU derivative.
     */
    fun leakyReluDerivative(x: Double, alpha: Double = 0.01): Double = if (x > 0) 1.0 else alpha

    /**
     * ELU (Exponential Linear Unit).
     */
    fun elu(x: Double, alpha: Double = 1.0): Double = if (x > 0) x else alpha * (exp(x) - 1)

    /**
     * SELU (Scaled ELU).
     */
    fun selu(x: Double): Double {
        val alpha = 1.6732632423543772848170429916717
        val scale = 1.0507009873554804934193349852946
        return scale * (if (x > 0) x else alpha * (exp(x) - 1))
    }

    /**
     * Swish activation: x * sigmoid(x)
     */
    fun swish(x: Double): Double = x * sigmoid(x)

    /**
     * GELU (Gaussian Error Linear Unit).
     */
    fun gelu(x: Double): Double {
        return 0.5 * x * (1 + kotlin.math.tanh(sqrt(2 / PI) * (x + 0.044715 * x.pow(3))))
    }

    /**
     * Softplus: log(1 + e^x)
     */
    fun softplus(x: Double): Double = ln(1 + exp(x))

    /**
     * Softsign: x / (1 + |x|)
     */
    fun softsign(x: Double): Double = x / (1 + abs(x))

    // ========== Vector Activation Functions ==========

    /**
     * Softmax activation for a vector.
     */
    fun softmax(x: List<Double>): List<Double> {
        val maxVal = x.maxOrNull() ?: return x
        val exps = x.map { exp(it - maxVal) }
        val sumExps = exps.sum()
        return exps.map { it / sumExps }
    }

    /**
     * Log-softmax for numerical stability.
     */
    fun logSoftmax(x: List<Double>): List<Double> {
        val maxVal = x.maxOrNull() ?: return x
        val shifted = x.map { it - maxVal }
        val logSumExp = ln(shifted.sumOf { exp(it) })
        return shifted.map { it - logSumExp }
    }

    // ========== Loss Functions ==========

    /**
     * Mean Squared Error.
     */
    fun mse(predicted: List<Double>, actual: List<Double>): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null
        val sum = predicted.zip(actual).sumOf { (p, a) -> (p - a).pow(2) }
        return sum / predicted.size
    }

    /**
     * Root Mean Squared Error.
     */
    fun rmse(predicted: List<Double>, actual: List<Double>): Double? {
        val mseVal = mse(predicted, actual) ?: return null
        return sqrt(mseVal)
    }

    /**
     * Mean Absolute Error.
     */
    fun mae(predicted: List<Double>, actual: List<Double>): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null
        val sum = predicted.zip(actual).sumOf { (p, a) -> abs(p - a) }
        return sum / predicted.size
    }

    /**
     * Binary Cross-Entropy loss.
     */
    fun binaryCrossEntropy(predicted: List<Double>, actual: List<Double>, epsilon: Double = 1e-15): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null

        val sum = predicted.zip(actual).sumOf { (p, a) ->
            val clippedP = p.coerceIn(epsilon, 1 - epsilon)
            -a * ln(clippedP) - (1 - a) * ln(1 - clippedP)
        }

        return sum / predicted.size
    }

    /**
     * Categorical Cross-Entropy loss.
     */
    fun categoricalCrossEntropy(predicted: List<Double>, actual: List<Double>, epsilon: Double = 1e-15): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null

        val sum = predicted.zip(actual).sumOf { (p, a) ->
            val clippedP = p.coerceIn(epsilon, 1 - epsilon)
            -a * ln(clippedP)
        }

        return sum
    }

    /**
     * Huber loss (smooth L1).
     */
    fun huberLoss(predicted: List<Double>, actual: List<Double>, delta: Double = 1.0): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null

        val sum = predicted.zip(actual).sumOf { (p, a) ->
            val error = abs(p - a)
            if (error <= delta) {
                0.5 * error.pow(2)
            } else {
                delta * error - 0.5 * delta.pow(2)
            }
        }

        return sum / predicted.size
    }

    // ========== Metrics ==========

    /**
     * Accuracy for classification.
     */
    fun accuracy(predicted: List<Int>, actual: List<Int>): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null
        val correct = predicted.zip(actual).count { (p, a) -> p == a }
        return correct.toDouble() / predicted.size
    }

    /**
     * Precision for binary classification.
     */
    fun precision(predicted: List<Int>, actual: List<Int>, positiveClass: Int = 1): Double? {
        if (predicted.size != actual.size) return null

        val truePositives = predicted.zip(actual).count { (p, a) -> p == positiveClass && a == positiveClass }
        val falsePositives = predicted.zip(actual).count { (p, a) -> p == positiveClass && a != positiveClass }

        val total = truePositives + falsePositives
        return if (total == 0) 0.0 else truePositives.toDouble() / total
    }

    /**
     * Recall for binary classification.
     */
    fun recall(predicted: List<Int>, actual: List<Int>, positiveClass: Int = 1): Double? {
        if (predicted.size != actual.size) return null

        val truePositives = predicted.zip(actual).count { (p, a) -> p == positiveClass && a == positiveClass }
        val falseNegatives = predicted.zip(actual).count { (p, a) -> p != positiveClass && a == positiveClass }

        val total = truePositives + falseNegatives
        return if (total == 0) 0.0 else truePositives.toDouble() / total
    }

    /**
     * F1 Score.
     */
    fun f1Score(predicted: List<Int>, actual: List<Int>, positiveClass: Int = 1): Double? {
        val p = precision(predicted, actual, positiveClass) ?: return null
        val r = recall(predicted, actual, positiveClass) ?: return null

        return if (p + r == 0.0) 0.0 else 2 * p * r / (p + r)
    }

    /**
     * R² (coefficient of determination).
     */
    fun r2Score(predicted: List<Double>, actual: List<Double>): Double? {
        if (predicted.size != actual.size || predicted.isEmpty()) return null

        val mean = actual.average()
        val ssRes = predicted.zip(actual).sumOf { (p, a) -> (a - p).pow(2) }
        val ssTot = actual.sumOf { (it - mean).pow(2) }

        return if (ssTot == 0.0) 1.0 else 1 - ssRes / ssTot
    }

    // ========== Normalization ==========

    /**
     * Min-max normalization to [0, 1].
     */
    fun minMaxNormalize(data: List<Double>): List<Double> {
        val minVal = data.minOrNull() ?: return data
        val maxVal = data.maxOrNull() ?: return data
        val range = maxVal - minVal

        return if (range == 0.0) {
            data.map { 0.5 }
        } else {
            data.map { (it - minVal) / range }
        }
    }

    /**
     * Z-score normalization (standardization).
     */
    fun zScoreNormalize(data: List<Double>): List<Double> {
        if (data.isEmpty()) return data

        val mean = data.average()
        val variance = data.sumOf { (it - mean).pow(2) } / data.size
        val std = sqrt(variance)

        return if (std == 0.0) {
            data.map { 0.0 }
        } else {
            data.map { (it - mean) / std }
        }
    }

    /**
     * L2 normalization (unit vector).
     */
    fun l2Normalize(data: List<Double>): List<Double> {
        val norm = sqrt(data.sumOf { it.pow(2) })
        return if (norm == 0.0) data else data.map { it / norm }
    }

    /**
     * Batch normalization.
     */
    fun batchNormalize(
        data: List<Double>,
        gamma: Double = 1.0,
        beta: Double = 0.0,
        epsilon: Double = 1e-5
    ): List<Double> {
        if (data.isEmpty()) return data

        val mean = data.average()
        val variance = data.sumOf { (it - mean).pow(2) } / data.size

        return data.map { x ->
            gamma * (x - mean) / sqrt(variance + epsilon) + beta
        }
    }

    // ========== Utility Functions ==========

    /**
     * One-hot encode a single integer.
     */
    fun oneHotEncode(value: Int, numClasses: Int): List<Double>? {
        if (value < 0 || value >= numClasses) return null
        return List(numClasses) { if (it == value) 1.0 else 0.0 }
    }

    /**
     * One-hot encode a list of integers.
     */
    fun oneHotEncode(values: List<Int>, numClasses: Int): List<List<Double>>? {
        return values.map { oneHotEncode(it, numClasses) ?: return null }
    }

    /**
     * Argmax: index of maximum value.
     */
    fun argmax(values: List<Double>): Int? {
        if (values.isEmpty()) return null
        return values.indices.maxByOrNull { values[it] }
    }

    /**
     * Argmin: index of minimum value.
     */
    fun argmin(values: List<Double>): Int? {
        if (values.isEmpty()) return null
        return values.indices.minByOrNull { values[it] }
    }

    /**
     * Clip values to a range.
     */
    fun clip(data: List<Double>, minVal: Double, maxVal: Double): List<Double> {
        return data.map { it.coerceIn(minVal, maxVal) }
    }

    /**
     * Cosine similarity between two vectors.
     */
    fun cosineSimilarity(a: List<Double>, b: List<Double>): Double? {
        if (a.size != b.size || a.isEmpty()) return null

        val dotProduct = a.zip(b).sumOf { (x, y) -> x * y }
        val normA = sqrt(a.sumOf { it.pow(2) })
        val normB = sqrt(b.sumOf { it.pow(2) })

        return if (normA == 0.0 || normB == 0.0) 0.0 else dotProduct / (normA * normB)
    }

    /**
     * Euclidean distance between two vectors.
     */
    fun euclideanDistance(a: List<Double>, b: List<Double>): Double? {
        if (a.size != b.size) return null
        return sqrt(a.zip(b).sumOf { (x, y) -> (x - y).pow(2) })
    }

    /**
     * Manhattan distance between two vectors.
     */
    fun manhattanDistance(a: List<Double>, b: List<Double>): Double? {
        if (a.size != b.size) return null
        return a.zip(b).sumOf { (x, y) -> abs(x - y) }
    }
}
